### PR TITLE
add support for nacl SealedBox and other extras

### PR DESCRIPTION
**Pull request by steverweber**
### What does this PR do?

https://github.com/saltstack/salt/issues/41831
a cleanup of PR: https://github.com/saltstack/salt/pull/41833 , https://github.com/saltstack/salt/pull/42323

### New Behavior

add support for nacl SealedBox

add support for enc and dec of files

### Tests written?

Yes

### CODE REVIEW

**Comment by steverweber** [comment]
need help with jenkins.

**Comment by rallytime** [comment]
Hi @steverweber It looks like the only related problem here is the lint failures:
https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/13034/violations/file/salt/modules/nacl.py/

If you hover over the red exclamation marks, it will tell you what the problem is. In this case, you need two spaces before the inline comments. Hope that helps. :)

File: salt/modules/nacl.py [hunk 1]
```diff
@@ -11,112 +11,157 @@
...
+
+    salt-run nacl.enc 'asecretpass'
+    salt-run nacl.dec 'tqXzeIJnTAM9Xf0mdLcpEdklMbfBGPj2oTKmlgrm3S1DTVVHNnh9h8mU1GKllGq/+cYsk6m5WhGdk58='
+
+.. code-block:: yam
+    # a salt developers minion could have pillar data that includes a nacl public key
+    nacl.config:
+        pk: '/kfGX7PbWeu099702PBbKWLpG/9p06IQRswkdWHCDk0='
+
+The developer can then use a less secure system to encrypt data.
```
Comment (cachedout) [inline salt/modules/nacl.py]: `less-secure`

File: salt/modules/nacl.py [hunk 2]
```diff
@@ -130,91 +175,286 @@ def _get_config(**kwargs):
...
+
+    if pk_file is None:
+        pk_file = '{0}.pub'.format(sk_file)
+
+    if sk_file and pk_file is None:
+        if not os.path.isfile(sk_file):
+            kp = libnacl.public.SecretKey()
+            with salt.utils.fopen(sk_file, 'w') as keyf:
+                keyf.write(base64.b64encode(kp.sk))
+            if not salt.utils.is_windows():
```
Comment (cachedout) [inline salt/modules/nacl.py]: Are we leaving this open to the world in Windows here then? We may need to account for this. cc: @twangboy and @damon-atkins for suggestions here.

File: salt/runners/nacl.py [hunk 3]
```diff
@@ -1,42 +1,167 @@
...
 
 .. code-block:: bash
 
-    salt-run nacl.enc 'data'
+    salt-call nacl.enc_file /tmp/cert.crt out=/tmp/cert.nacl
+    # or more advanced
+    cert=$(cat /tmp/cert.crt)
+    salt-call --out=newline_values_only nacl.enc_pub data="$cert" > /tmp/cert.nacl
+
+In Pillars rended with jinja be sure to include `|json` so line breaks are encoded:
```
Comment (cachedout) [inline salt/runners/nacl.py]: `pillars`

File: salt/runners/nacl.py [hunk 4]
```diff
@@ -1,42 +1,167 @@
...
+
+.. code-block:: jinja
+
+    cert: "{{salt.nacl.dec('S2uogToXkgENz9...085KYt')|json}}"
+
+In States rendered with jinja it is also good pratice to include `|json`:
```
Comment (cachedout) [inline salt/runners/nacl.py]: `states`

File: salt/modules/nacl.py (hunk 2 above)
Comment (cachedout) [inline salt/modules/nacl.py]: @steverweber Did you have any feedback here?

Comment (steverweber) [inline salt/modules/nacl.py]: nope.

Comment (steverweber) [inline salt/modules/nacl.py]: how does a salt master protect it's files in windows? does it?

**Comment by steverweber** [comment]
@cachedout we have two other users with PR that are waiting for this to merge.
can we merge soon?

thanks.

https://github.com/saltstack/salt/pull/41956
https://github.com/saltstack/salt/pull/41868

File: salt/modules/nacl.py (hunk 2 above)
Comment (cachedout) [inline salt/modules/nacl.py]: @twangboy ?

Comment (twangboy) [inline salt/modules/nacl.py]: @cachedout Maybe if it is Windows:
```
import salt.utils.win_functions
import salt.utils.win_dacl
cur_user = salt.utils.win_functions.get_current_user()
salt.utils.win_dacl.set_owner(sk_file, cur_user)
salt.utils.win_dacl.set_permissions(sk_file, cur_user, 'full_control', 'grant', reset_perms=True, protected=True)
```
That will make the user running the command the owner and remove all permissions to the file except for the current user. The current user will have full control permissions.

Comment (steverweber) [inline salt/modules/nacl.py]: tomorrow I'll update code to include the above

**Comment by cachedout** [comment]
@steverweber I am just waiting on the Windows changed discussed in the review comments and then we can get this in.

**Review by twangboy — APPROVED** [review]

**Pull request merged**